File: NEsper/pattern/EvalObserverStateNode.cs
using System;

using net.esper.pattern.observer;

using org.apache.commons.logging;

namespace net.esper.pattern
{
    /// <summary>
	/// This class represents the state of an eventObserver sub-expression
	/// in the evaluation state tree.
	/// </summary>
    public sealed class EvalObserverStateNode
		: EvalStateNode
		, ObserverEventEvaluator
    {
        private EventObserver eventObserver;

        /// <summary> Constructor.</summary>
        /// <param name="parentNode">is the parent evaluator to call to indicate truth value
        /// </param>
        /// <param name="evalObserverNode">the factory node associated to the state
        /// </param>
        /// <param name="beginState">contains the events that make up prior matches
        /// </param>
        /// <param name="context">contains handles to services required
        /// </param>
        public EvalObserverStateNode(Evaluator parentNode, 
									 EvalObserverNode evalObserverNode,
									 MatchedEventMap beginState,
									 PatternContext context)
            : base(evalObserverNode, parentNode, null)
        {
            if (log.IsDebugEnabled)
            {
                log.Debug(".constructor");
            }

            eventObserver = evalObserverNode.ObserverFactory.MakeObserver(context, beginState, this, null, null);
        }

        /// <summary>
        /// Indicate an event for evaluation (sub-expression the observer represents has turned true).
        /// </summary>
        /// <param name="matchEvent">is the matched events so far</param>
        public void ObserverEvaluateTrue(MatchedEventMap matchEvent)
        {
            this.ParentEvaluator.EvaluateTrue(matchEvent, this, true);
        }

        /// <summary>
        /// Indicate that the observer turned permanently false.
        /// </summary>
        public void ObserverEvaluateFalse()
        {
            this.ParentEvaluator.EvaluateFalse(this);
        }

        /// <summary>
        /// Starts the event expression or an instance of it.
        /// Child classes are expected to initialize and Start any event listeners
        /// or schedule any time-based callbacks as needed.
        /// </summary>
        public override void Start()
        {
            eventObserver.StartObserve();
        }

        /// <summary>
        /// Stops the event expression or an instance of it. Child classes are expected to free resources
        /// and Stop any event listeners or remove any time-based callbacks.
        /// </summary>
        public override void Quit()
        {
            eventObserver.StopObserve();
        }

        /// <summary>
        /// Accept a visitor. Child classes are expected to invoke the visit method on the visitor instance
        /// passed in.
        /// </summary>
        /// <param name="visitor">on which the visit method is invoked by each node</param>
        /// <param name="data">any additional data the visitor may need is passed in this parameter</param>
        /// <returns>
        /// any additional data the visitor may need or null
        /// </returns>
        public override Object Accept(EvalStateNodeVisitor visitor, Object data)
        {
            return visitor.Visit(this, data);
        }

        /// <summary>
        /// Pass the visitor to all child nodes.
        /// </summary>
        /// <param name="visitor">is the instance to be passed to all child nodes</param>
        /// <param name="data">any additional data the visitor may need is passed in this parameter</param>
        /// <returns>
        /// any additional data the visitor may need or null
        /// </returns>
        public override Object ChildrenAccept(EvalStateNodeVisitor visitor, Object data)
        {
            return data;
        }

        /// <summary>
        /// Returns a <see cref="T:System.String"></see> that represents the current <see cref="T:System.Object"></see>.
        /// </summary>
        /// <returns>
        /// A <see cref="T:System.String"></see> that represents the current <see cref="T:System.Object"></see>.
        /// </returns>
        public override String ToString()
        {
            return "EvalObserverStateNode eventObserver=" + eventObserver;
        }

        private static readonly Log log = LogFactory.GetLog(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
    }
}

File: trunk/trunk/esper/src/main/java/net/esper/pattern/EvalObserverStateNode.java
package net.esper.pattern;

import net.esper.pattern.observer.EventObserver;
import net.esper.pattern.observer.ObserverEventEvaluator;
import net.esper.pattern.observer.ObserverFactory;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;


/**
 * This class represents the state of an eventObserver sub-expression in the evaluation state tree.
 */
public final class EvalObserverStateNode extends EvalStateNode implements ObserverEventEvaluator
{
    private EventObserver eventObserver;

    /**
     * Constructor.
     * @param parentNode is the parent evaluator to call to indicate truth value
     * @param observerFactory is the observer factory that makes observer instances
     * @param beginState contains the events that make up prior matches
     * @param context contains handles to services required
     */
    public EvalObserverStateNode(Evaluator parentNode,
                             ObserverFactory observerFactory,
                                   MatchedEventMap beginState,
                                   PatternContext context)
    {
        super(parentNode);

        if (log.isDebugEnabled())
        {
            log.debug(".constructor");
        }

        eventObserver = observerFactory.makeObserver(context, beginState, this);
    }

    public void observerEvaluateTrue(MatchedEventMap matchEvent)
    {
        this.getParentEvaluator().evaluateTrue(matchEvent, this, true);
    }

    public void observerEvaluateFalse()
    {
        this.getParentEvaluator().evaluateFalse(this);
    }

    public final void start()
    {
        eventObserver.startObserve();
    }

    protected final void quit()
    {
        eventObserver.stopObserve();
    }

    public final Object accept(EvalStateNodeVisitor visitor, Object data)
    {
        return visitor.visit(this, data);
    }

    public final Object childrenAccept(EvalStateNodeVisitor visitor, Object data)
    {
        return data;
    }

    public final String toString()
    {
        return "EvalObserverStateNode eventObserver=" + eventObserver;
    }

    private static final Log log = LogFactory.getLog(EvalObserverStateNode.class);
}
